compiler/luci/pass/src/PropagateQParamBackwardPass.cpp

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2022 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/PropagateQParamBackwardPass.h"
+
+#include <luci/IR/CircleNodes.h>
+#include <luci/IR/CircleNodeVisitor.h>
+#include <luci/Log.h>
+
+namespace
+{
+
+// Visitor to propagate quantization parameters backwards
+struct PropagateQParamBackward final : public luci::CircleNodeMutableVisitor<void>
+{
+  PropagateQParamBackward(loco::DataType output) : _output_type(output) {}
+
+private:
+  loco::DataType _output_type;
+
+  void visit(luci::CircleNode *) {}
+};
+
+} // namespace
+
+namespace luci
+{
+
+bool PropagateQParamBackwardPass::run(loco::Graph *g)
+{
+  LOGGER(l);
+
+  // We use post-order traversal as qparam is propagated backward
```

Comment:
If `qparam` is propagated backward, I think `qparam` is propagated with `op3 --> op2 --> op1` sequence, am I right?
```
[op1] --- [op2] --- [op3] -- output
```

If so, I think pre-order traversal is correct... isn't it?

Reply:
Ah, you're right. `loco::postorder_traversal` is in fact post-order traversal of args, which is the opposite of data flow direction. Thanks.

Reply:
> If so, I think pre-order traversal is correct... isn't it?

Hmm.. actually, reverse post-order traversal (topological order) would be right. Pre-ordering makes overwrite of qparam when there is a diamond-shaped dataflow. (See https://en.wikipedia.org/wiki/Depth-first_search#Vertex_orderings for more details)

Reply:
I think both pre-order and reverse(post-order) is fine when diamond-shaped dataflow :)
But, for diamond-shaped flow, both of them would be problem..!

```
[op1] -+-- [op2] --+- [op3] --- output
       +-- [op4] --+
```

For above example graph, `reverse(loco::postorder_traversal)` would return `reverse(op1, op2, op4, op3) = (op3, op4, op2, op1)`.
Then `op2` overwrites qparam which is propagated from `op4` :(

In this case, what is intended behavior?

Reply:
> In this case, what is intended behavior?

It depends on whether `op1` is const or not
1. If `op1` is const : `op1` is cloned and quantized for each Op (`op2` and `op4`)
2. If `op1` is not const : qparam of `op2` and `op4` is not propagated to `op1`, because `op1` has multiple successors.

> both pre-order and reverse(post-order) is fine when diamond-shaped dataflow :)

You're right. I was worried about the mismatch of qparam in the overwritten op (`op1`'s ofm and its ifms), but the source op (`op4`) also has the same problem.

Both traversals (pre-order and reverse post-order) would be ok, because qparam will not be propagated if `op1` is not const. In this case, I'll go with reverse post-order for ease of implementation.

Reply:
> I was worried about the mismatch of qparam in the overwritten op (op1's ofm and its ifms), but the source op (op4) would also have the same problem.

Then it would be better to add some validation/assertion code to check it later :)

> I'll go with reverse post-order for ease of implementation.

I agree with this because pre-order traversal is not implemented in `loco`.
So using `postorder_traversal` seems much safer and better :D